### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by jicata**
closed Implement C# Unit Testing Strategy SoftUni-Internal/suls-issues/issues/1780

### CODE REVIEW

File: Open Judge System/Data/OJS.Data/Migrations/DefaultMigrationConfiguration.cs [hunk 1]
```diff
@@ -313,6 +314,18 @@ protected void SeedSubmissionTypes(OjsDbContext context)
                     AllowedFileExtensions = "zip",
                     AllowBinaryFilesUpload = true,
                 },
+                new SubmissionType
+                {
+                    Name = "C# Unit Tests Tester",
+                    CompilerType = CompilerType.None,
+                    AdditionalCompilerArguments = "--noresult --inprocess",
+                    ExecutionStrategyType =
```
Comment (rozay) [inline Open Judge System/Data/OJS.Data/Migrations/DefaultMigrationConfiguration.cs]: Make `ExecutionStrategyType = ExecutionStrategyType.CSharpUnitTestsRunnerExecutionStrategy,` on one line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        private const string TestedCode = "TestedCode.cs";
+        private const string CsProjFileSearchPattern = "*.csproj";
+        private const string ProjFileExtenstion = ".csproj";
+        private const string DllFileExtension = ".dll";
+
+        public CSharpUnitTestsRunnerExecutionStrategy(string nUnitConsoleRunnerPath)
+        {
+            this.NUnitConsoleRunnerPath = nUnitConsoleRunnerPath;
+            this.WorkingDirectory = DirectoryHelpers.CreateTempDirectory();
+
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Remove unnecessary blank line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            var checker = Checker.CreateChecker(
+                executionContext.CheckerAssemblyName,
+                executionContext.CheckerTypeName,
+                executionContext.CheckerParameter);
+
+            result = this.RunUnitTests(executionContext, executor, checker, result, csProjFilePath);
+            return result;
+        }
+
+        private ExecutionResult RunUnitTests(ExecutionContext executionContext, IExecutor executor, IChecker checker, ExecutionResult result, string csProjFilePath)
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Place each parameter on separate line in order to not exceed 120 characters convention limit.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                        break;
+                    case ProcessExecutionResultType.MemoryLimit:
+                        testResult.ResultType = TestRunResultType.MemoryLimit;
+                        break;
+                    case ProcessExecutionResultType.Success:
+                        int totalTests = 0;
+                        int passedTests = 0;
+                 
+                        this.ExtractTestResult(processExecutionResult.ReceivedOutput, ref passedTests, ref totalTests);
+                        string message = "Test Passed!";
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+                                message = "Not all tests passed on the correct solution.";
+                            }
+                        }
+
+                        var checkerResult = checker.Check(test.Input, message, test.Output, test.IsTrialTest);
+                        testResult.ResultType = checkerResult.IsCorrect ? TestRunResultType.CorrectAnswer : TestRunResultType.WrongAnswer;
+                        testResult.CheckerDetails = checkerResult.CheckerDetails;
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(processExecutionResult), "Invalid ProcessExecutionResultType value.");
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Place parameters on separate lines in order to not exceed 120 characters convention limit.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            Dictionary<string, string> nUnitMetaData = new Dictionary<string, string>();
+            nUnitMetaData.Add("Private", "True");
+            project.AddItem(
+                "Reference",
+                "nunit.framework, Version=3.6.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL",
+                nUnitMetaData);
+    
+            // If we use NUnit we don't really need the VSTT, it will save us copying of the .dll
+            var vsTestFrameworkReference =
+                project.Items.FirstOrDefault(
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: According the convention `project.Items` should be on the upper line, `.FirstOrDefault(x =>` on new one and the rest on new line too.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+            // Remove previous NUnit reference (the path is probably pointing to the users package folder)
+            var nUnitPrevReference = project.Items.FirstOrDefault(x => x.EvaluatedInclude.Contains("nunit.framework"));
+            if (nUnitPrevReference != null)
+            {
+                project.RemoveItem(nUnitPrevReference);
+            }
+
+            // Add our NUnit Reference, if private is false, the .dll will not be copied and the tests will not run
+            Dictionary<string, string> nUnitMetaData = new Dictionary<string, string>();
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                        {
+                            originalTestsPassed = passedTests;
+                            if (totalTests != passedTests)
+                            {
+                                message = "Not all tests passed on the correct solution.";
+                            }
+                        }
+
+                        var checkerResult = checker.Check(test.Input, message, test.Output, test.IsTrialTest);
+                        testResult.ResultType = checkerResult.IsCorrect ? TestRunResultType.CorrectAnswer : TestRunResultType.WrongAnswer;
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Place `? TestRunResultType.CorrectAnswer` on new line and `: TestRunResultType.WrongAnswer;` right below.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 9]
```diff
@@ -206,17 +209,18 @@ private void CorrectProjectReferences(Project project)
...
             project.AddItem(
                 "Reference",
                 "nunit.framework, Version=3.6.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL",
                 nUnitMetaData);
     
             // If we use NUnit we don't really need the VSTT, it will save us copying of the .dll
-            var vsTestFrameworkReference =
-                project.Items.FirstOrDefault(
-                    x => x.EvaluatedInclude.Contains("Microsoft.VisualStudio.QualityTools.UnitTestFramework"));
+            var vsTestFrameworkReference = project.Items.
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Place '.' on the line below

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 10]
```diff
@@ -206,17 +209,18 @@ private void CorrectProjectReferences(Project project)
...
+                FirstOrDefault(x => 
+                x.EvaluatedInclude.Contains("Microsoft.VisualStudio.QualityTools.UnitTestFramework"));
```
Comment (rozay) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Add `tab`

File: Open Judge System/Workers/OJS.Workers.LocalWorker/Settings.cs [hunk 11]
```diff
@@ -18,6 +18,10 @@ static Settings()
 
         public static string CPlusPlusGccCompilerPath => GetSetting("CPlusPlusGccCompilerPath");
 
+        public static string NUnitFrameworkFrameworkPath => GetSetting("NUnitFrameworkPath");
```
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/Settings.cs]: Remove the NUnitFrameworkPath setting as it is not required.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    using System.IO;
+    using System.Linq;
+    using System.Text.RegularExpressions;
+
+    using Microsoft.Build.Evaluation;
+
+    using OJS.Common.Extensions;
+    using OJS.Common.Models;
+    using OJS.Workers.Checkers;
+    using Common;
```
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Add the full path to the resource (ex. OJS.Common)

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 13]
```diff
@@ -4,14 +4,14 @@
...
-    using Common;
+    using System.Text.RegularExpressions;   
     using Microsoft.Build.Evaluation;
 
-    using OJS.Common.Extensions;
-    using OJS.Common.Models;
     using OJS.Workers.Checkers;
     using OJS.Workers.Executors;
+    using OJS.Workers.Common;
+    using OJS.Common.Extensions;
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: The usings in this block must be ordered alphabetically (OJS.Common.... must be before OJS.Workers...)

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        {
+            ExecutionResult result = new ExecutionResult();
+            byte[] userSubmissionContent = executionContext.FileContent;
+
+            var submissionFilePath = $"{this.WorkingDirectory}\\{ZippedSubmissionName}";
+            File.WriteAllBytes(submissionFilePath, userSubmissionContent);
+            FileHelpers.UnzipFile(submissionFilePath, this.WorkingDirectory);
+            File.Delete(submissionFilePath);
+
+            string csProjFilePath = FileHelpers.FindFirstFileMatchingPattern(
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: you may use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+                this.WorkingDirectory,
+                CsProjFileSearchPattern);
+
+            // Edit References in Project file
+            var project = new Project(csProjFilePath);
+            this.CorrectProjectReferences(project);
+            project.Save(csProjFilePath);
+            project.ProjectCollection.UnloadAllProjects();
+
+            // Initially set isCompiledSucessfully to true
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: the comment is not necessary IMO

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+            var submissionFilePath = $"{this.WorkingDirectory}\\{ZippedSubmissionName}";
+            File.WriteAllBytes(submissionFilePath, userSubmissionContent);
+            FileHelpers.UnzipFile(submissionFilePath, this.WorkingDirectory);
+            File.Delete(submissionFilePath);
+
+            string csProjFilePath = FileHelpers.FindFirstFileMatchingPattern(
+                this.WorkingDirectory,
+                CsProjFileSearchPattern);
+
+            // Edit References in Project file
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: Is comment useful enough? Shall we keep it?

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+                // Run unit tests on the resulting .dll
+                var processExecutionResult = executor.Execute(
+                    this.NUnitConsoleRunnerPath,
+                    string.Empty,
+                    executionContext.TimeLimit,
+                    executionContext.MemoryLimit,
+                    arguments);
+
+                // Construct and figure out what the Test result is
+                TestResult testResult = new TestResult()
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+                result.TestResults.Add(testResult);
+                count++;
+            }
+
+            return result;
+        }
+
+        private void ExtractTestResult(string receivedOutput, ref int passedTests, ref int totalTests)
+        {
+            Regex testResultsRegex =
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/CSharpUnitTestsRunnerExecutionStrategy.cs]: use `var`

**Pull request merged**